### PR TITLE
[GROW-1717] /api/search/providers 

### PR DESCRIPTION
**Pull request by thiskevinwang**
JIRA | [GROW-1717](https://capsulerx.atlassian.net/browse/GROW-1717)
:--- | :---

# Description
## Related
https://github.com/CapsuleHealth/consumer-web/pull/1458

### types 
This installs several types dependencies

tbd...

### CODE REVIEW

File: src/app/routes/searchRoutes/index.ts [hunk 1]
```diff
@@ -1,25 +1,24 @@
 import express, { Request, Response, NextFunction } from 'express';
 import * as Sentry from '@sentry/node';
+import * as mockData from './mockData.json';
 
 const searchRouter = express.Router();
 
-searchRouter.get('/provider/:query?', async (req: Request, res: Response, next: NextFunction) => {
+searchRouter.get('/providers', async (req: Request, res: Response, next: NextFunction) => {
```
Comment (klein-thibault) [inline src/app/routes/searchRoutes/index.ts]: Do you usually code everything inside the router? I used to create controller classes that the router would call, just to separate responsibilities between different files.

File: src/app/routes/searchRoutes/searchRoutes.test.ts [hunk 2]
```diff
@@ -12,8 +13,8 @@ describe('GET /api/search/provider', function() {
 
   it('responds with empty array', function() {
     return request(app)
-      .get('/api/search/provider?query=Dr+Susan&location=07302')
+      .get('/api/search/providers?query=Dr+Susan&location=07302')
       .expect(200)
-      .expect([])
+      .expect(mockData.data);
```
Comment (klein-thibault) [inline src/app/routes/searchRoutes/searchRoutes.test.ts]: Have you considered snapshot testing? It would allow you to easily make sure the whole response is valid.

File: swagger.yaml [hunk 3]
```diff
@@ -16,3 +25,14 @@ paths:
         - url: 'http://localhost:3005'
     servers:
       - url: 'http://localhost:3005'
+
+  /api/search/providers:
+    get:
+      description: Returns a list of providers
+      responses:
```
Comment (klein-thibault) [inline swagger.yaml]: This one is only a mock for now but are you planning on documenting all the different errors possible?

File: src/app/routes/searchRoutes/index.ts (hunk 1 above)
Comment (thiskevinwang) [reply src/app/routes/searchRoutes/index.ts]: Do you have a quick example of this?

(This is some of the first Express work I'm doing, so it's new territory to me... show me your ways!!!)

File: swagger.yaml (hunk 3 above)
Comment (thiskevinwang) [reply swagger.yaml]: Yes of course, I want to document everything!

Swagger is also new to me... (I previously only know self-documenting, machine-readable _schemas_ 🤣)

File: src/app/routes/searchRoutes/searchRoutes.test.ts (hunk 2 above)
Comment (thiskevinwang) [reply src/app/routes/searchRoutes/searchRoutes.test.ts]: I wanted to do this but our API key for [Ribbon](https://ribbon.readme.io/docs/search-for-providers1) expired, and I didn't previously save any Postman requests ☹️

The next best thing was the example "JSON" on their page ^, which isn't ideal.

File: swagger.yaml [hunk 4]
```diff
@@ -2,9 +2,18 @@ openapi: 3.0.1
 info:
   title: Consumer Web API Documentation
   description: Documents end points for the Capsule Consumer Web BFF
-  version: '0.1'
+  version: '0.2.0'
+
 servers:
   - url: 'http://localhost:3005'
+    description: Local development endpoint
+  - url: 'FIXME Sandbox'
```
Comment (alexbwell) [inline swagger.yaml]: You can get the URLs here: https://capsulerx.atlassian.net/browse/GROW-1687

File: swagger.yaml (hunk 3 above)
Comment (alexbwell) [reply swagger.yaml]: This endpoint should handle all errors that get thrown from the 3rd party call and still return a 200. Should we explicitly document the "standard" errors like network errors, etc.?

File: swagger.yaml [hunk 5]
```diff
@@ -16,3 +25,14 @@ paths:
         - url: 'http://localhost:3005'
     servers:
       - url: 'http://localhost:3005'
+
+  /api/search/providers:
+    get:
+      description: Returns a list of providers
+      responses:
+        '200':
+          description: Providers
```
Comment (alexbwell) [inline swagger.yaml]: Should describe this endpoint's _search functionality_ in more detail.

File: src/app/routes/searchRoutes/index.ts [hunk 6]
```diff
@@ -1,25 +1,24 @@
...
       // res.json({name: null, phone: null, address: null, specialty: null});
-      
     } else {
       res.send(response);
-      
+
       //placeholder
-      const providerSearch = "Provider Search";
+      const providerSearch = 'Provider Search';
       Sentry.captureMessage(`${providerSearch} responded with a non-200 status code: ${response.status}`);
```
Comment (alexbwell) [inline src/app/routes/searchRoutes/index.ts]: This should still return a 200 status with empty result set.

File: src/app/routes/searchRoutes/index.ts [hunk 7]
```diff
@@ -1,25 +1,24 @@
...
     // const response = {status: 200, data: {query: req.query.query, location: req.query.location}};
-    const response = {status: 200, data: []};
+    const response = { status: 200, data: mockData };
 
     if (response.status === 200) {
-
       //placeholder
-      res.send(response.data)
+      res.send(response.data.data);
       // res.json({name: null, phone: null, address: null, specialty: null});
```
Comment (alexbwell) [inline src/app/routes/searchRoutes/index.ts]: Don't need this anymore💥

File: src/app/routes/searchRoutes/index.ts [hunk 8]
```diff
@@ -1,25 +1,24 @@
...
 
 const searchRouter = express.Router();
 
-searchRouter.get('/provider/:query?', async (req: Request, res: Response, next: NextFunction) => {
+searchRouter.get('/providers', async (req: Request, res: Response, next: NextFunction) => {
   try {
     // This is a placeholder for a response from a third party provider database.
     // const response = {status: 200, data: {query: req.query.query, location: req.query.location}};
-    const response = {status: 200, data: []};
+    const response = { status: 200, data: mockData };
```
Comment (alexbwell) [inline src/app/routes/searchRoutes/index.ts]: Can we be a little "extra" here and actually fake a search by filtering the mock data on the search term? It'll help test more edge cases in QA.

File: src/app/routes/searchRoutes/index.ts (hunk 1 above)
Comment (klein-thibault) [reply src/app/routes/searchRoutes/index.ts]: Correct, I would have Router/Controller/Model files and the flow would be Router -> Controller -> Model then back.

I'd recommend sticking to the same pattern everywhere though, so you keep things consistent. Even if the controller is tiny I would still create it and test it separately. That way, when it grows, you already have the structure in place. And it helps new engineers to grasp the architecture everwhere they go.

File: src/app/routes/searchRoutes/searchRoutes.test.ts (hunk 2 above)
Comment (klein-thibault) [reply src/app/routes/searchRoutes/searchRoutes.test.ts]: Not sure I understand your comment, FYI I was talking about [Jest snapshot](https://jestjs.io/docs/en/snapshot-testing) testing technique.

Sounds like you were talking about mocking the request to Ribbon? In that case, I used [nock](https://github.com/nock/nock) which is really awesome for mocking 3rd party APIs.

File: swagger.yaml (hunk 3 above)
Comment (klein-thibault) [reply swagger.yaml]: > This endpoint should handle all errors that get thrown from the 3rd party call and still return a 200. Should we explicitly document the "standard" errors like network errors, etc.?

Definitely should have all the generic errors defined as reused errors that you add here. For the 3rd party call errors, Idk what the plan is to handle them but in Swagger, we're only interested in the errors that front-end will consume. If the 3rd party returns errors that are caused by the front-end, they should be documented here. If 3rd party failed for an unknown reason (if they are down for instance) maybe return a 500 server error?

File: src/app/routes/searchRoutes/searchRoutes.test.ts (hunk 2 above)
Comment (alexbwell) [reply src/app/routes/searchRoutes/searchRoutes.test.ts]: The code doesn't make a "live" request for data at the moment, so I'd hold off on Nock this time and we'll add it as part of our integration with Ribbon API ticket.

File: swagger.yaml (hunk 3 above)
Comment (alexbwell) [reply swagger.yaml]: The product spec requires us to return an empty result set if we receive an error from the API. The only errors that can be returned to client here should be network errors that aren't really part of this endpoint's behaviour...

**Comment by thiskevinwang** [comment]
Going to close this and rework the project structure to follow @klein-thibault's suggestion of "separating concerns"!
